#**Titanic - Machine Learning from Disaster**

*   Utilizaremos os datasets disponíveis no [Kaggle Competitions](https://www.kaggle.com/competitions/titanic/overview);
*   A pontuação é a porcentagem de sobreviventes previstos corretamente (Acurácia);


# Parte 1  
# **Compreendendo e manipulando as bases de dados**

**Com o auxílio da biblioteca [ydata-profiling](https://github.com/ydataai/ydata-profiling), vamos obter um resumo da base de dados para análise:**

In [ ]:
#Importando pandas
import pandas as pd

In [ ]:
#Visualizando base de treino
treino = pd.read_csv("train.csv")
treino.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [ ]:
#Instalando ydata_profiling e importando o ProfileReport
!pip install ydata_profiling -q
from ydata_profiling import ProfileReport

#Gerando visualização
profile = ProfileReport(treino, title = 'Titanic_Treino')

#Criando relatório
profile.to_file("Titanic_Treino.html")

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 357.9/357.9 kB 5.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 104.8/104.8 kB 8.5 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 686.1/686.1 kB 10.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 293.3/293.3 kB 11.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 296.5/296.5 kB 11.6 MB/s eta 0:00:00


/usr/local/lib/python3.10/dist-packages/ydata_profiling/profile_report.py:354: UserWarning: Try running command: 'pip install --upgrade Pillow' to avoid ValueError
  warnings.warn(


Summarize dataset:   0%|          | 0/5 [00:00<?, ?it/s]

Generate report structure:   0%|          | 0/1 [00:00<?, ?it/s]

Render HTML:   0%|          | 0/1 [00:00<?, ?it/s]

Export report to file:   0%|          | 0/1 [00:00<?, ?it/s]

**Analisando a base de treino:**

In [ ]:
#Verificando informações da base
treino.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [ ]:
#Verificando tipos de dados
treino.dtypes.value_counts()

int64      5
object     5
float64    2
Name: count, dtype: int64

In [ ]:
#Verificando valores nulos
treino.isnull().sum().sort_values(ascending = False).head()

Cabin          687
Age            177
Embarked         2
PassengerId      0
Survived         0
dtype: int64

*   Existem valores vazios na coluna **Age**, isso não faz sentido, bem como na coluna **Embarked**, necessário tratamento de dados na base de treino

In [ ]:
#Analisando formato da base
treino.shape

(891, 12)

*   Existem muitos valores vazios em Cabin, **687 das 891 linhas estão vazias**

**Analisando a base de teste:**

In [ ]:
#Visualizando base de teste
teste = pd.read_csv("test.csv")
teste.head()

PassengerId  Pclass                                          Name     Sex  \
0          892       3                              Kelly, Mr. James    male   
1          893       3              Wilkes, Mrs. James (Ellen Needs)  female   
2          894       2                     Myles, Mr. Thomas Francis    male   
3          895       3                              Wirz, Mr. Albert    male   
4          896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female   

    Age  SibSp  Parch   Ticket     Fare Cabin Embarked  
0  34.5      0      0   330911   7.8292   NaN        Q  
1  47.0      1      0   363272   7.0000   NaN        S  
2  62.0      0      0   240276   9.6875   NaN        Q  
3  27.0      0      0   315154   8.6625   NaN        S  
4  22.0      1      1  3101298  12.2875   NaN        S

In [ ]:
#Verificando informações da base
teste.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 418 entries, 0 to 417
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  418 non-null    int64  
 1   Pclass       418 non-null    int64  
 2   Name         418 non-null    object 
 3   Sex          418 non-null    object 
 4   Age          332 non-null    float64
 5   SibSp        418 non-null    int64  
 6   Parch        418 non-null    int64  
 7   Ticket       418 non-null    object 
 8   Fare         417 non-null    float64
 9   Cabin        91 non-null     object 
 10  Embarked     418 non-null    object 
dtypes: float64(2), int64(4), object(5)
memory usage: 36.0+ KB


In [ ]:
#Verificando valores nulos
teste.isnull().sum().sort_values(ascending = False).head()

Cabin          327
Age             86
Fare             1
PassengerId      0
Pclass           0
dtype: int64

*   Existem colunas com valores vazios em **lugares diferentes das bases**, ou
seja, será necessário tratamento de dados na base de teste, além da base treino, para que os modelos possam trabalhar nas duas bases.

In [ ]:
#Verificando Cardinalidade dos dados
treino.nunique().sort_values(ascending=False)

PassengerId    891
Name           891
Ticket         681
Fare           248
Cabin          147
Age             88
SibSp            7
Parch            7
Pclass           3
Embarked         3
Survived         2
Sex              2
dtype: int64

*   Vemos que há grande variedade de dados nas colunas **Ticket, Cabin e Nome**. Isso prejudica a previsão, alta cardinaidade em colunas de texto torna o modelo menos generalizavel.

**Tratando dados vazios:**
*   Base treino: Age e Embarked
*   Base teste: Age e Fare

In [ ]:
#Eliminando colunas para base de treino
treino = treino.drop(['Name','Ticket','Cabin'], axis = 1)

In [ ]:
#Eliminando colunas para base de teste
teste = teste.drop(['Name','Ticket','Cabin'], axis = 1)

In [ ]:
#Verificando a média das idades na base de treino
treino.Age.mean()

29.69911764705882

In [ ]:
#Substituindo idades vazias pela média das idades
treino.loc[treino.Age.isnull(), 'Age'] = treino.Age.mean()
teste.loc[teste.Age.isnull(), 'Age'] = teste.Age.mean()

In [ ]:
#Verificando valores da coluna Embarked
treino.Embarked.value_counts()

Embarked
S    644
C    168
Q     77
Name: count, dtype: int64

In [ ]:
#Selecionando a moda (valor com maior representatividade)
treino.Embarked.mode()[0]

'S'

In [ ]:
#Substituindo vazios pela moda
treino.loc[treino.Embarked.isnull(), 'Embarked'] = treino.Embarked.mode()[0]

In [ ]:
#Verificando a média da coluna Fare na base de teste
teste.Fare.mean()

35.627188489208635

In [ ]:
#Substituindo vazioss pela média da Fare
teste.loc[teste.Fare.isnull(), 'Fare'] = teste.Fare.mean()

In [ ]:
#Nova verificação de valores vazios
treino.isnull().sum().sort_values(ascending = False).head()

PassengerId    0
Survived       0
Pclass         0
Sex            0
Age            0
dtype: int64

In [ ]:
#Nova verificação de valores vazios
teste.isnull().sum().sort_values(ascending = False).head()

PassengerId    0
Pclass         0
Sex            0
Age            0
SibSp          0
dtype: int64

**Manipulando somente as colunas com valores numéricos:**

In [ ]:
#Verificando colunas de texto
treino.columns[treino.dtypes != 'object']

Index(['PassengerId', 'Survived', 'Pclass', 'Age', 'SibSp', 'Parch', 'Fare'], dtype='object')

In [ ]:
teste.columns[teste.dtypes != 'object']

Index(['PassengerId', 'Pclass', 'Age', 'SibSp', 'Parch', 'Fare'], dtype='object')

In [ ]:
#Atribuindo colunas numéricas à variáveis
col_treino_num = treino.columns[treino.dtypes != 'object']
col_teste_num = teste.columns[teste.dtypes != 'object']

In [ ]:
#Selecionando somente as colunas com valores numéricos
treino_num = treino.loc[:,col_treino_num]
teste_num = teste.loc[:,col_teste_num]

**Criação de modelos para classificação:**
*   1 - Precisamos dividir a base de treino em treino e validação com o [train_test_split](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html);
*   2 - Vamos usar os seguintes classificadores:
    * [Árvore de Classificação](https://scikit-learn.org/stable/modules/generated/sklearn.tree.DecisionTreeClassifier.html);
    * [K Vizinhos mais Próximos](https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.KNeighborsClassifier.html);
    * [Regressão Logística](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html);

In [ ]:
#Importar train_test_split
from sklearn.model_selection import train_test_split

In [ ]:
#Separando base de treino em X e Y
X = treino_num.drop(['PassengerId','Survived'],axis=1)
y = treino.Survived

*   *   **X** = a base completa, excluindo a coluna que precisa ser prevista;
        *   Retiramos a coluna **PassengerID** para que não atrapalhe na previsão, não eliminamos ela da base pois será necessária no momento de submissão ao Kaggle.
    *   **y** = coluna que precisa ser prevista;


In [ ]:
#Separando em treino e validação
X_train, X_val, y_train, y_val = train_test_split(X,y,test_size=0.33,random_state=42)

In [ ]:
#Árvore de classificação:

#Importação
from sklearn import tree

#Criando cassificador
clf_ac = tree.DecisionTreeClassifier(random_state=42)

#Fit com os dados
clf_ac = clf_ac.fit(X_train, y_train)

#Realizando previsão
y_pred_ac = clf_ac.predict(X_val)

In [ ]:
#K Vizinhos mais próximos:

#Importação
from sklearn.neighbors import KNeighborsClassifier

#Criando cassificador
clf_knn = KNeighborsClassifier(n_neighbors=3)

#Fit com os dados
clf_knn = clf_knn.fit(X_train, y_train)

#Realizando previsão

y_pred_knn = clf_knn.predict(X_val)

In [ ]:
#Regressão logística

#Importação
from sklearn.linear_model import LogisticRegression

#Criando cassificador
clf_rl = LogisticRegression(random_state=42)

#Fit com os dados
clf_rl = clf_rl.fit(X_train, y_train)

#Realizando previsão
y_pred_rl = clf_rl.predict(X_val)

**Avaliando os modelos por:**
*   1 - [Acurácia](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.accuracy_score.html)
*   2 - [Matriz de confusão](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.confusion_matrix.html)

In [ ]:
#Importação Acurácia:
from sklearn.metrics import accuracy_score

In [ ]:
#Árvore de classificação
accuracy_score(y_val, y_pred_ac)

0.6169491525423729

In [ ]:
#K Vizinhos mais próximos
accuracy_score(y_val, y_pred_knn)

0.6610169491525424

In [ ]:
#Regressão Logística
accuracy_score(y_val, y_pred_rl)

0.7254237288135593

In [ ]:
#Importação Matriz de Confusão
from sklearn.metrics import confusion_matrix

In [ ]:
#Árvore de classificação
confusion_matrix(y_val, y_pred_ac)

array([[125,  50],
       [ 63,  57]])

In [ ]:
#K Vizinhos mais próximos
confusion_matrix(y_val, y_pred_knn)

array([[133,  42],
       [ 58,  62]])

In [ ]:
#Regressão Logística
confusion_matrix(y_val, y_pred_rl)

array([[156,  19],
       [ 62,  58]])

**Fazendo previsão na base teste:**
*   Usaremos o modelo com melhor precisão para fazer a previsão (predict)

In [ ]:
#Visualizando o X_train
X_train.head()

Pclass        Age  SibSp  Parch     Fare
6         1  54.000000      0      0  51.8625
718       3  29.699118      0      0  15.5000
685       2  25.000000      1      2  41.5792
73        3  26.000000      1      0  14.4542
882       3  22.000000      0      0  10.5167

In [ ]:
##Visualizando somente as colunas com valores numéricos da base teste
teste_num.head()

PassengerId  Pclass   Age  SibSp  Parch     Fare
0          892       3  34.5      0      0   7.8292
1          893       3  47.0      1      0   7.0000
2          894       2  62.0      0      0   9.6875
3          895       3  27.0      0      0   8.6625
4          896       3  22.0      1      1  12.2875

In [ ]:
#Eliminar coluna ID para que a base teste fique igual a base treino
X_teste = teste_num.drop(['PassengerId'], axis=1)

In [ ]:
#Aplicando regrassão logística
y_pred = clf_rl.predict(X_teste)

In [ ]:
#Criando coluna com os resultados da previsão na base teste
teste['Survived'] = y_pred

In [ ]:
#Selecionando colunas de interesse para fazer envio ao Kaggle
envio_kaggle = teste[['PassengerId','Survived']]

In [ ]:
#Exportando .csv
envio_kaggle.to_csv('resultados_parte01.csv',index=False)

Resultado da submissão: **0.66746**